### PR TITLE
Implement a first version of a stream wrapper

### LINKED ISSUE #1: More detailed general architecture for decryption of file content
Here is a more specific architecture for decrypting file content and making it flexible within Drupal.
- Implement a route like `encrypt.files` which allows code to link to our encrypted file URL.
  In this URL we could encode the profile, so something like `/encrypt/{key_profile_name}/{path}`
- In the controller for that route, use `encrypt://{key_profile_name}/{path}` as URL, but keep mostly the existing implementation of `\Drupal\system\FileDownloadController::download` to keep everything in the scope of Drupal
-  stream wrapper that looks like : `encrypt://{key_profile}/path/...`
  This calls out to the decrypt method itself:
  - The decryption supports streaming: Stream in the file content via the decryption.
  - Decryption doesn't support streaming, load the file and decrpyt it via ->decrypt directly, basically exactly how http://cgit.drupalcode.org/encrypted_files/tree/includes/EncryptedStreamWrapper.inc is dealing with it.
## A streaming interface

In order to make streaming as easy as possible one could use basically `https://github.com/paragonie/halite/blob/master/src/Contract/StreamInterface.php` as interface for description streaming.

What do you think about this more low level details?

### PR DESCRIPTION
**Pull request by dawehner**
Tries to resolve #1

### CODE REVIEW

File: src/EncryptStreamWrapper.php [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+   *
+   * @return \Drupal\Core\Entity\EntityInterface|\Drupal\encrypt\EncryptionProfileInterface|null
+   */
+  protected function getEncryptionProfile($encryption_profile) {
+    /** @var \Drupal\encrypt\EncryptionProfileManager $profile_manager */
+    $profile_manager = \Drupal::service('encrypt.encryption_profile.manager');
+    return $profile_manager->getEncryptionProfile($encryption_profile);
+  }
+
+  protected function extractEncryptionProfile($uri) {
```
Comment (nerdstein) [inline src/EncryptStreamWrapper.php]: Missing comment

File: src/EncryptStreamWrapper.php [hunk 2]
```diff
@@ -0,0 +1,193 @@
+<?php
+
```
Comment (nerdstein) [inline src/EncryptStreamWrapper.php:2]: Can we add some comments here introducing the class?

File: src/EncryptStreamWrapper.php [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+   *
+   * Note that this static method is used by \Drupal\system\Form\FileSystemForm
+   * so you should alter that form or substitute a different form if you change
+   * the class providing the stream_wrapper.private service.
+   *
+   * @return string
+   *   The base path for private://.
+   */
+  public static function basePath() {
+    return Settings::get('encrypted_file_path', '');
```
Comment (nerdstein) [inline src/EncryptStreamWrapper.php:87]: Is this configurable? What happens if someone changes this after files are uploaded?

File: src/EncryptStreamWrapper.php [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+  /**
+   * Returns the base path for encrypted://.
+   *
+   * Note that this static method is used by \Drupal\system\Form\FileSystemForm
+   * so you should alter that form or substitute a different form if you change
+   * the class providing the stream_wrapper.private service.
+   *
+   * @return string
+   *   The base path for private://.
```
Comment (nerdstein) [inline src/EncryptStreamWrapper.php]: s/private/encrypted/r

**Comment by nerdstein** [comment]
This looks good to me. I left some small comments

File: src/EncryptStreamWrapper.php [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    // don't resave the file.
+    if ((strpos($this->mode, 'r') !== FALSE) &&
+      (strpos($this->mode, '+') === FALSE)) {
+      fclose($this->handle);
+      return;
+    }
+    // Encrypt file and save.
+    rewind($this->handle);
+    $file_contents = stream_get_contents($this->handle);
+    file_put_contents($this->getLocalPath(), $this->encrypt($file_contents, $this->encryptionProfile));
```
Comment (nerdstein) [inline src/EncryptStreamWrapper.php]: Just wondering, this seems to suggest this writes out the encrypted file upon close. Can it/should it write before hand using the streaming algorithm?

File: src/EncryptStreamWrapper.php (hunk 2 above)
Comment (dawehner) [inline src/EncryptStreamWrapper.php:2]: Sure, good point!

File: src/EncryptStreamWrapper.php (hunk 3 above)
Comment (dawehner) [inline src/EncryptStreamWrapper.php:87]: For me this is a similar setting to 

```
$settings['file_private_path'] = '';
$settings['file_public_path'] = 'sites/default/files';
```

so something which is environment specific, for which `settings.php` is a good location.

Do you think having this in actual configuration would be better?

File: src/EncryptStreamWrapper.php (hunk 5 above)
Comment (dawehner) [inline src/EncryptStreamWrapper.php]: Yeah this honestly doesn't solve yet the partially more complex usecase of a streaming algorithm.

Comment (nerdstein) [inline src/EncryptStreamWrapper.php]: Cool, just checking.

**Pull request merged**